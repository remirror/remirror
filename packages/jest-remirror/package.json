{
  "name": "jest-remirror",
  "version": "3.0.0-beta.6",
  "description": "A text editor for react built with prosemirror",
  "keywords": [
    "prosemirror",
    "react",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-remirror.js",
      "require": "./dist/jest-remirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-remirror-environment.js",
      "require": "./environment/dist/jest-remirror-environment.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/jest-remirror.cjs",
  "module": "./dist/jest-remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.6",
    "@remirror/pm": "3.0.0-beta.4",
    "@remirror/preset-core": "3.0.0-beta.6",
    "@testing-library/dom": "^9.3.0",
    "@types/sanitize-html": "^2.9.0",
    "jest-prosemirror": "3.0.0-beta.4",
    "sanitize-html": "^2.10.0",
    "tiny-warning": "^1.0.3"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "@types/node": "^16.18.34",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "@types/node": "*",
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    },
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-remirror.d.ts",
        "import": "./dist/jest-remirror.js",
        "require": "./dist/jest-remirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-remirror-environment.d.ts",
        "import": "./environment/dist/jest-remirror-environment.js",
        "require": "./environment/dist/jest-remirror-environment.cjs"
      },
      "./package.json": "./package.json"
    },
    "main": "./dist/jest-remirror.cjs",
    "module": "./dist/jest-remirror.js",
    "type": "module",
    "types": "./dist/jest-remirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts"
    ]
  }
}
