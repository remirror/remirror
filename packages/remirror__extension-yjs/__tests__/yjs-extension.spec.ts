import { jest } from '@jest/globals';
import { extensionValidityTest, renderEditor } from 'jest-remirror';
import { hideConsoleError } from 'testing';
import { yUndoPluginKey } from 'y-prosemirror';
import { WebrtcProvider } from 'y-webrtc';
import { Doc } from 'yjs';

import { defaultDestroyProvider, YjsExtension, YjsOptions } from '../';

let provider = new WebrtcProvider('global', new Doc());

beforeEach(() => {
  provider.destroy();
  provider = new WebrtcProvider('global', new Doc());
});

extensionValidityTest(YjsExtension, {
  getProvider: () => provider,
});

describe('configuration', () => {
  hideConsoleError(true);

  it('throws without a provider', () => {
    expect(() => renderEditor<YjsExtension>([new YjsExtension()])).toThrowErrorMatchingSnapshot();
  });

  it('calls the destroy provider', () => {
    const destroyProvider: any = jest.fn(defaultDestroyProvider);
    const { manager } = create({ destroyProvider });

    manager.destroy();
    expect(destroyProvider).toHaveBeenCalledTimes(1);

    // Make sure the method is only called once.
    manager.destroy();
    expect(destroyProvider).toHaveBeenCalledTimes(1);
  });

  it('can update options dynamically', () => {
    const { manager } = create();
    const extension = manager.getExtension(YjsExtension);

    const getProvider: any = jest.fn(() => provider);
    const destroyProvider: any = jest.fn(defaultDestroyProvider);

    extension.setOptions({ getProvider, destroyProvider });
    expect(extension.options.getProvider).toBe(getProvider);
    expect(extension.options.destroyProvider).toBe(destroyProvider);
    expect(destroyProvider).not.toHaveBeenCalled();
    expect(getProvider).toHaveBeenCalledTimes(1);

    extension.setOptions({ getProvider: () => provider });
    expect(extension.options.destroyProvider).toHaveBeenCalledTimes(1);
  });

  it.each([{ protectedNodes: new Set<string>() }, { trackedOrigins: [] }, { disableUndo: true }])(
    'throws when updating undo-related options (%o)',
    (option: any) => {
      const { manager } = create();
      const extension = manager.getExtension(YjsExtension);

      expect(() => extension.setOptions({ ...option })).toThrowErrorMatchingSnapshot();
    },
  );

  it('uses the same undo manager in each state', () => {
    const { manager } = create();

    const state = manager.createState();
    const initialUndoManager = yUndoPluginKey.getState(state).undoManager;

    const secondState = manager.createState();
    const secondUndoManager = yUndoPluginKey.getState(secondState).undoManager;
    expect(secondUndoManager).toBe(initialUndoManager);
  });
});

describe('commands', () => {
  hideConsoleError(true);

  it('can undo', () => {
    const { nodes, add } = create();
    const { p, doc } = nodes;
    add(doc(p('<cursor>')))
      .insertText('Text goes here')
      .callback(({ commands }) => {
        expect(commands.yUndo.enabled()).toBeTrue();
        commands.yUndo();
        expect(commands.yUndo.enabled()).toBeFalse();
      })
      .callback((content) => {
        expect(content.state.doc).toEqualRemirrorDocument(doc(p('')));
      });
  });

  it('can redo', () => {
    const { nodes, add } = create();
    const { p, doc } = nodes;

    add(doc(p('<cursor>')))
      .insertText('Text goes here')
      .callback(({ commands }) => {
        expect(commands.yRedo.enabled()).toBeFalse();
        commands.yUndo();
        expect(commands.yRedo.enabled()).toBeTrue();
        commands.yRedo();
        expect(commands.yRedo.enabled()).toBeFalse();
      })
      .callback(({ state }) => {
        expect(state.doc).toEqualRemirrorDocument(doc(p('Text goes here')));
      });
  });
});

describe('keymap', () => {
  hideConsoleError(true);

  it('can undo', () => {
    const { nodes, add } = create();
    const { p, doc } = nodes;

    add(doc(p('<cursor>')))
      .insertText('Text goes here')
      .shortcut('Mod-z')
      .callback((content) => {
        expect(content.state.doc).toEqualProsemirrorNode(doc(p('')));
      });
  });

  it('can redo', () => {
    const { nodes, add } = create();
    const { p, doc } = nodes;

    add(doc(p('<cursor>')))
      .insertText('Text goes here')
      .shortcut('Mod-z')
      .shortcut('Mod-y')
      .callback((content) => {
        expect(content.state.doc).toEqualProsemirrorNode(doc(p('Text goes here')));
      });
  });
});

/**
 * Create a `yjs` extension.
 */
function create(options?: Partial<YjsOptions>, excludeHistory = true) {
  let excludeExtensions: ['history'] | undefined;

  if (excludeHistory) {
    excludeExtensions = ['history'];
  }

  const extension = new YjsExtension({
    getProvider: () => provider,
    ...options,
  });

  return renderEditor([extension], { core: { excludeExtensions } });
}
