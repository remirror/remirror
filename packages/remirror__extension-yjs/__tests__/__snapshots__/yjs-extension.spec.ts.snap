// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configuration throws when updating undo-related options ({ disableUndo: true }) 1`] = `
"A call to \`extension.setOptions\` was made with invalid keys.

Invalid properties passed into the 'setOptions()' method: [\\"disableUndo\\"].

For more information visit https://remirror.io/docs/errors#rmr0103"
`;

exports[`configuration throws when updating undo-related options ({ protectedNodes: Set(0) {} }) 1`] = `
"A call to \`extension.setOptions\` was made with invalid keys.

Invalid properties passed into the 'setOptions()' method: [\\"protectedNodes\\"].

For more information visit https://remirror.io/docs/errors#rmr0103"
`;

exports[`configuration throws when updating undo-related options ({ trackedOrigins: [ [length]: 0 ] }) 1`] = `
"A call to \`extension.setOptions\` was made with invalid keys.

Invalid properties passed into the 'setOptions()' method: [\\"trackedOrigins\\"].

For more information visit https://remirror.io/docs/errors#rmr0103"
`;

exports[`configuration throws without a provider 1`] = `
"An error occurred within an extension. More details should be made available.

You must provide a YJS Provider to the \`YjsExtension\`.

For more information visit https://remirror.io/docs/errors#rmr0100"
`;
